workers/data_refinery_workers/processors/create_compendia.py

```diff
@@ -0,0 +1,225 @@
+     - "Reset" zero values that were set to NA in RNA-seq samples (i.e., make these zero again) in combined_matrix
+     - Transpose combined_matrix; transposed_matrix
+     - Perform imputation of missing values with IterativeSVD (rank=10) on the transposed_matrix; imputed_matrix
+     - Untranspose imputed_matrix (genes are now rows, samples are now columns)
+     - Quantile normalize imputed_matrix where genes are rows and samples are columns
+
+    """
+    job_context['time_start'] = timezone.now()
+
+    # Combine all microarray samples with a full join to form a microarray_expression_matrix (this may end up being a DataFrame)
+    microarray_expression_matrix = job_context['microarray_inputs']
+
+    # Combine all RNA-seq samples (lengthScaledTPM) with a full outer join to form a rnaseq_expression_matrix
+    rnaseq_expression_matrix = job_context['rnaseq_inputs']
+
+    # Calculate the sum of the lengthScaledTPM values for each row (gene) of the rnaseq_expression_matrix (rnaseq_row_sums)
+    rnaseq_row_sums = np.sum(rnaseq_expression_matrix, axis=1)
+
+    # Calculate the 10th percentile of rnaseq_row_sums
+    rnaseq_tenth_percentile = np.percentile(rnaseq_row_sums, 10)
+
+    # Drop all rows in rnaseq_expression_matrix with a row sum < 10th percentile of rnaseq_row_sums; this is now filtered_rnaseq_matrix
+    # TODO: This is probably a better way to do this with `np.where`
+    rows_to_filter = []
+    for (x, sum_val) in rnaseq_row_sums.items():
+        if sum_val < rnaseq_tenth_percentile:
+            rows_to_filter.append(x)
+
+    filtered_rnaseq_matrix = rnaseq_expression_matrix.drop(rows_to_filter)
+
+    # log2(x + 1) transform filtered_rnaseq_matrix; this is now log2_rnaseq_matrix
+    filtered_rnaseq_matrix_plus_one = filtered_rnaseq_matrix + 1
+    log2_rnaseq_matrix = np.log2(filtered_rnaseq_matrix_plus_one)
+
+    # Set all zero values in log2_rnaseq_matrix to NA, but make sure to keep track of where these zeroes are
+    log2_rnaseq_matrix[log2_rnaseq_matrix==0]=np.nan
+
+    # Perform a full outer join of microarray_expression_matrix and log2_rnaseq_matrix; combined_matrix
+    combined_matrix = pd.merge(microarray_expression_matrix, log2_rnaseq_matrix, how='outer', left_index=True, right_index=True)
+
+    # Remove genes (rows) with >30% missing values in combined_matrix
+    thresh = combined_matrix.shape[1] * .7
+    row_filtered_combined_matrix = combined_matrix.dropna(axis=0, thresh=thresh, how='any')
+
+    # Remove samples (columns) with >50% missing values in combined_matrix
+    # XXX: Find better test data for this!
+    thresh = len(combined_matrix.columns) * .5
+    row_col_filtered_combined_matrix_samples = row_filtered_combined_matrix.dropna(axis=1, thresh=thresh, how='any')
+
+    # "Reset" zero values that were set to NA in RNA-seq samples (i.e., make these zero again) in combined_matrix
+    # XXX: IterativeSVD doesn't see 0.000 as something to impute, only NA.
+    # So, we skip this step!
```

Comment:
We don't want to skip this step. As we talked about yesterday, zeroes in RNA-seq data could be technical or biological (we have no easy way of knowing, here). We're setting the zeroes to NAs above to make it easier to perform the row and column filtering in lines 125-132. 

We do not want to impute the values that were originally zero in RNA-seq data, so we want to set them to `0.000` instead of leaving them as `NA`. 

Because the indices are changing based on the filtering, I don't think it will be sufficient to keep track of the indices where the zeros _were_. Instead, it makes sense to me to keep track of the gene labels and sample labels for values that were originally zero in the RNA-seq data and then set (any remaining post-filtering) values to zero.

You can also do this _after_ imputation but before quantile normalization, but I can see an argument where it's more efficient to do it _before_ imputation (= less values to impute) but I have no idea whether this actually makes a difference performance-wise.

Happy to chat in person or poke around at this step.